viewing_party/party.py

```diff
@@ -1,23 +1,155 @@
+import statistics
+from statistics import mode
+from frozendict import frozendict
+
+
 # ------------- WAVE 1 --------------------
 
-def create_movie(title, genre, rating):
-    pass
+def create_movie(movie_title, genre, rating):
+    if movie_title and genre and rating:
+        return {
+            "title": movie_title,
+            "genre": genre,
+            "rating": rating
+        }
+    
+    return None
+
+# user data is a dict and it has a key "watched" --> list 
+# add movie to this list 
+
+def add_to_watched(user_data, movie):
+    user_data["watched"].append(movie)
+    return user_data
+
+
+def add_to_watchlist(user_data, movie):
+    user_data["watchlist"].append(movie)
+    return user_data
+
+def watch_movie(user_data, title):
+    for movie in user_data["watchlist"]:
+        if movie["title"] == title:
+            user_data["watchlist"].remove(movie)
+            user_data["watched"].append(movie)
+            return user_data
+        
+    return user_data
+    
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
+def get_watched_avg_rating(user_data):
+    if not user_data["watched"]:
+        return 0
+    
+    running_total = 0
+    for movie in user_data["watched"]:
+        running_total += movie["rating"]
+
+    return running_total / len(user_data["watched"])
+
+
+def get_most_watched_genre(user_data):
+    if not user_data["watched"]:
+        return None
+    
+    genre_list = []
+    for movie in user_data["watched"]:
+        genre_list.append(movie["genre"])
+
+    return mode(genre_list)
 
+    
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
-# -----------------------------------------
+
+def get_titles_user_watched(user_data):
+    movies_user_watched = set()
+
+    for movie in user_data["watched"]:
+        movies_user_watched.add(movie["title"])
+
+    return movies_user_watched
+
+
+def get_titles_friends_watched(user_data):
+    movies_friends_have_watched = set()
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            movies_friends_have_watched.add(movie["title"])
+
+    return movies_friends_have_watched
+
+
+def get_unique_watched(user_data):
+    titles_friends_have_watched = get_titles_friends_watched(user_data)
+
+    movies_only_user_watched = []
+    for movie in user_data["watched"]:
+        if movie["title"] not in titles_friends_have_watched:
+            movies_only_user_watched.append(movie)
+
+
+    return movies_only_user_watched
+
+
+
+def get_friends_unique_watched(user_data):
+    titles_user_watched = get_titles_user_watched(user_data)
+
+    movies_only_friends_watched = []
+
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie["title"] not in titles_user_watched:

```

Comment:
There is a way to do this without frozendict. You can do this:
```
if movie["title"] not in titles_user_watched and movie no in movies_only_friends_watched`
```
This will remove the duplicates that the set removes for you.

Reply:
This was such a helpful review, thank you very much, Nancy! Appreciate the alternative to the frozen structures!